src/ds/thread_messaging.h

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/logger.h"
+#ifdef USE_MPSCQ
+#  include "ds/mpscq.h"
+#endif
+#include "ds/ts.h"
+
+#include <atomic>
+#include <cstddef>
+
+namespace enclave
+{
+  struct thread_msg
+  {
+    void (*cb)(std::unique_ptr<thread_msg>);
+    std::atomic<thread_msg*> next = nullptr;
+    uint64_t padding[14];
+  };
+
+  template <typename Tpayload>
+  struct tmsg
+  {
+    tmsg(void (*_cb)(std::unique_ptr<tmsg<Tpayload>>)) :
+      cb(reinterpret_cast<void (*)(std::unique_ptr<thread_msg>)>(_cb)),
+      next(nullptr)
+    {
+      CheckInvariants();
```

Comment:
For consistency with the CCF code, these should be `ThreadMsg`, `typename Payload`, `TMsg`, `check_invariants`.